sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -114,8 +114,41 @@ case class ConcatWs(children: Seq[Expression])
         boolean ${ev.isNull} = ${ev.primitive} == null;
       """
     } else {
-      // Contains a mix of strings and array<string>s. Fall back to interpreted mode for now.
-      super.genCode(ctx, ev)
+      val list = ctx.freshName("list")
+      val array = ctx.freshName("array")
+
+      val sep = children.head.gen(ctx)
+      val vararg = children.tail.map { child =>
+        val eval = child.gen(ctx)
+        child.dataType match {
+          case StringType =>
+            s"""
+              ${eval.code}
+              $list.add(${eval.isNull} ? (UTF8String) null : ${eval.primitive});
+             """
+          case _: ArrayType =>
+            val size = ctx.freshName("n")
+            val j = ctx.freshName("j")
+            s"""
+              ${eval.code}
+              if (!${eval.isNull}) {
+                final int $size = ${eval.primitive}.numElements();
+                for (int $j = 0; $j < $size; $j ++) {
```

Comment:
Minor: `j` is only available in loop scope, so we don't need to create a unique name for it.